File: src/main/java/org/usfirst/irs1318/gamesim/engine/event/EventLog.java
package org.usfirst.irs1318.gamesim.engine.event;

import java.time.Instant;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public final class EventLog {
    private final List<Entry> entries = Collections.emptyList();

    public void addEntry(Entry entry) {
        this.entries.add(entry);
    }

    public static final class Entry {
        private final Instant instant;
        private final Map<String, String> content;

        public Entry(Instant instant, Map<String, String> content) {
            this.instant = instant;
            this.content = content;
        }

        public Instant getInstant() {
            return instant;
        }

        public Map<String, String> getContent() {
            return content;
        }

        @Override
        public String toString() {
            return "Entry{" +
                    "instant=" + instant +
                    ", content=" + content +
                    '}';
        }
    }
}

File: src/main/scala/org/usfirst/irs1318/gamesim/engine/event/EventLog.scala
package org.usfirst.irs1318.gamesim.engine.event

case class EventLog(entries: List[EventLog.Entry]) {
  @inline def mapEntries(fn: List[EventLog.Entry] => List[EventLog.Entry]): EventLog = copy(entries = fn(entries))

  def +(entry: EventLog.Entry): EventLog = mapEntries(_ :+ entry)

  def ++(eventLog: EventLog): EventLog = mapEntries(_ ++ eventLog.entries)
}

object EventLog {
  def empty: EventLog = EventLog(List.empty)

  case class Entry(content: Map[String, String]) {
    @inline def mapContent(fn: Map[String, String] => Map[String, String]): Entry = copy(content = fn(content))
  }

  object Entry {
    case class Builder(entry: Entry) {
      def this() = this(Entry(Map.empty))
      @inline def mapEntry(fn: Entry => Entry): Builder = copy(entry = fn(entry))
      def setContent(content: Map[String, String]): Builder = mapEntry(_.mapContent(_ => content))
      def putContent(name: String, value: String): Builder = mapEntry(_.mapContent(_ + ((name, value))))
      def build: Entry = entry
    }
  }
}
